src/tensors/tensor.jl

```diff
@@ -70,89 +78,216 @@ storagetype(::Type{<:TensorMap{T, S, N₁, N₂, A}}) where {T, S, N₁, N₂, A
+2) `data` can be an `AbstractMatrix` of size `(dim(codomain), dim(domain))`
+3) `data` can be an `AbstractArray` of rank `N₁ + N₂` with a size matching that of the domain
+   and codomain spaces, i.e. `size(data) == (dims(codomain)..., dims(domain)...)`
+
+In cases 2 and 3, the `TensorMap` constructor will reconstruct the tensor data such that the
+resulting tensor `t` satisfies `data == convert(Array, t)`, up to an error specified by `tol`.
+For the case where `sectortype(S) == Trivial` and `data isa DenseArray`, the `data` array is
+simply reshaped into a vector and used as in case 1 so that the memory will still be shared.
+In other cases, new memory will be allocated.
+
+Note that in the case of `N₁ + N₂ = 1`, case 3 also amounts to `data` being a vector, whereas
+when `N₁ + N₂ == 2`, case 2 and case 3 both require `data` to be a matrix. Such ambiguous cases
+are resolved by checking the size of `data` in an attempt to support all possible
+cases.
 
-# constructor starting from vector = independent data (N₁ + N₂ = 1 is special cased below)
-# documentation is captured by the case where `data` is a general array
-# here, we force the `T` argument to distinguish it from the more general constructor below
-function TensorMap{T}(
-        data::A, V::TensorMapSpace{S, N₁, N₂}
-    ) where {T, S, N₁, N₂, A <: DenseVector{T}}
-    return TensorMap{T, S, N₁, N₂, A}(data, V)
+!!! note
+    This constructor for case 2 and 3 only works for `sectortype` values for which conversion
+    to a plain array is possible, and only in the case where the `data` actually respects
+    the specified symmetry structure, up to a tolerance `tol`.
+"""
+function TensorMap(data::AbstractArray, V::TensorMapSpace; tol = sqrt(eps(real(float(eltype(data))))))
+    T = eltype(data)
+    A = _tensormap_storagetype(typeof(data))
+    return TensorMapWithStorage{T, A}(data, V; tol)
 end
-function TensorMap{T}(
-        data::DenseVector{T}, codomain::TensorSpace{S}, domain::TensorSpace{S}
-    ) where {T, S}
-    return TensorMap(data, codomain ← domain)
+TensorMap(data::AbstractArray, codom::TensorSpace, dom::TensorSpace; kwargs...) =
+    TensorMap(data, codom ← dom; kwargs...)
+Tensor(data::AbstractArray, codom::TensorSpace; kwargs...) =
+    TensorMap(data, codom ← one(codom); kwargs...)
+
+function TensorMapWithStorage{T, A}(
+        data::AbstractArray, V::TensorMapSpace; tol = sqrt(eps(real(float(eltype(data)))))
+    ) where {T, A}
+    # refer to specific raw data constructors if input is a vector of the correct length
+    ndims(data) == 1 && length(data) == dim(V) &&
+        return tensormaptype(spacetype(V), numout(V), numin(V), A)(data, V)
+
+    # special case trivial: refer to same method, but now with vector argument
+    sectortype(V) === Trivial &&
+        return tensormaptype(spacetype(V), numout(V), numin(V), A)(reshape(data, length(data)), V)
+
+    # do projection
+    t = TensorMapWithStorage{T, A}(undef, V)
+    t = project_symmetric!(t, data)
```

Comment:
just a note that this will not work with GPU-backed arrays until we write a `project_symmetric` which supports them. For those arrays, we'll currently need to do this projection in CPU memory and then copy over, but this overall shouldn't impact performance too much.

Reply:
I guess for now we should then specialize `project_symmetric!` in the extensions to go through CPU and go from there?

Reply:
We might also have an annoying problem with converting a `CuTensorMap` to a `CuArray`. I might even suggest refactoring the entire " `project_symmetric!` and check " routine into a separate function? Then `CuTensorMap` (or `ROCTensorMap`) can handle copying back and forth to/from CPU. But that could be done in a separate PR